cpp/examples/1billion/brc_pipeline.cpp

```diff
@@ -0,0 +1,172 @@
+    // process each byte range assigned to this thread
+    for (auto& br : byte_ranges) {
+      auto const input_table = load_chunk(input_file, br.first, br.second, stream);
+      auto const read_rows   = input_table->num_rows();
+      if (read_rows == 0) continue;
+
+      auto const cities = input_table->view().column(0);
+      auto const temps  = input_table->view().column(1);
+
+      std::vector<std::unique_ptr<cudf::groupby_aggregation>> aggregations;
+      aggregations.emplace_back(cudf::make_min_aggregation<cudf::groupby_aggregation>());
+      aggregations.emplace_back(cudf::make_max_aggregation<cudf::groupby_aggregation>());
+      aggregations.emplace_back(cudf::make_sum_aggregation<cudf::groupby_aggregation>());
+      aggregations.emplace_back(cudf::make_count_aggregation<cudf::groupby_aggregation>());
+      auto result = compute_results(cities, temps, std::move(aggregations), stream);
+
+      agg_data.emplace_back(
+        cudf::sort_by_key(result->view(), result->view().select({0}), {}, {}, stream));
+    }
+    // done with this stream
+    stream.synchronize_no_throw();
+  }
+};
+
+int main(int argc, char const** argv)
+{
+  if (argc < 2) {
+    std::cout << "required parameter: input-file-path\n";
+    std::cout << "optional parameters: chunk-count thread-count\n";
+    return 1;
+  }
+
+  auto const input_file   = std::string{argv[1]};
+  auto const divider      = (argc < 3) ? 25 : std::stoi(std::string(argv[2]));
+  auto const thread_count = (argc < 4) ? 2 : std::stoi(std::string(argv[3]));
+
+  std::cout << "input:   " << input_file << std::endl;
+  std::cout << "chunks:  " << divider << std::endl;
+  std::cout << "threads: " << thread_count << std::endl;
+
+  auto const mr_name = std::string("pool");  // "cuda"
+  auto resource      = create_memory_resource(mr_name);
+  auto smr = rmm::mr::statistics_resource_adaptor<rmm::mr::device_memory_resource>(resource.get());
+  rmm::mr::set_current_device_resource(&smr);
+  auto stream = cudf::get_default_stream();
+
+  std::filesystem::path p = input_file;
+  auto const file_size    = std::filesystem::file_size(p);
+
+  auto start = std::chrono::steady_clock::now();
+
+  std::size_t chunk_size = file_size / divider + ((file_size % divider) != 0);
+  std::size_t start_pos  = 0;
+
+  auto stream_pool = rmm::cuda_stream_pool(thread_count);
+  std::vector<std::vector<result_t>> chunk_results(thread_count);
+
+  std::vector<chunk_fn> chunks;
+  for (auto& cr : chunk_results) {
+    chunks.emplace_back(chunk_fn{input_file, cr, stream_pool.get_stream()});
+  }
+  for (std::size_t i = 0; i < divider; ++i) {
+    auto start = i * chunk_size;
+    auto size  = std::min(chunk_size, file_size - start);
+    chunks[i % thread_count].add_range(start, size);
+  }
+  std::vector<std::thread> threads;
+  for (auto& c : chunks) {
+    threads.emplace_back(std::thread{c});
+  }
+  for (auto& t : threads) {
+    t.join();
+  }
+
+  // in case some APIs are still running on the default stream
```

Comment:
Maybe kernels instead of APIs? The word API seems more like an abstract contract, not a unit of "real" work.
```suggestion
  // in case some kernels are still running on the default stream
```